server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -1322,20 +1811,41 @@ protected void doClose() {
             if (snapshots != null) {
                 int changedCount = 0;
                 final List<SnapshotsInProgress.Entry> entries = new ArrayList<>();
+                final Map<String, Set<ShardId>> reusedShardIdsByRepo = new HashMap<>();
                 for (SnapshotsInProgress.Entry entry : snapshots.entries()) {
                     ImmutableOpenMap.Builder<ShardId, ShardSnapshotStatus> shards = ImmutableOpenMap.builder();
                     boolean updated = false;
 
                     for (UpdateIndexShardSnapshotStatusRequest updateSnapshotState : tasks) {
+                        final ShardId finishedShardId = updateSnapshotState.shardId();
                         if (entry.snapshot().equals(updateSnapshotState.snapshot())) {
                             logger.trace("[{}] Updating shard [{}] with status [{}]", updateSnapshotState.snapshot(),
-                                updateSnapshotState.shardId(), updateSnapshotState.status().state());
+                                    finishedShardId, updateSnapshotState.status().state());
                             if (updated == false) {
                                 shards.putAll(entry.shards());
                                 updated = true;
                             }
-                            shards.put(updateSnapshotState.shardId(), updateSnapshotState.status());
+                            shards.put(finishedShardId, updateSnapshotState.status());
                             changedCount++;
+                        } else {
+                            final Set<ShardId> reusedShardIds =
+                                    reusedShardIdsByRepo.computeIfAbsent(entry.repository(), k -> new HashSet<>());
+                            if (entry.state().completed() == false && reusedShardIds.contains(finishedShardId) == false
```

Comment:
Does this logic still make sure that a snapshot that's taken contains a set of docs that existed sometime between start and end of snapshot command? Or will this batch up snapshots that had been running prior to the command?

Reply:
> Does this logic still make sure that a snapshot that's taken contains a set of docs that existed sometime between start and end of snapshot command? 

Yes. There is no smart logic here. We simply reassign the finished shards as (new `INIT` state shard snapshots) if there's other outstanding snapshots for them and put them into `reusedShardIds` to ensure we only reassign once per shard+repo => a new snapshot will be taken from a fresh index commit so the guarantees about the timing are unchanged relative to what we have right now.